
  <!DOCTYPE html>
<html lang="en" data-accent-color="violet" data-content_root="">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>Models - mango 0.2.1 documentation</title>
    <link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Bibliography" href="../bib.html" /><link rel="prev" title="Clients" href="clients.html" />

  <script>
    function setColorMode(t){let e=document.documentElement;e.setAttribute("data-color-mode",t);let a=window.matchMedia&&window.matchMedia("(prefers-color-scheme: dark)").matches,s=t;"auto"===t&&(s=a?"dark":"light"),"light"===s?(e.classList.remove("dark"),e.classList.add("light")):(e.classList.remove("light"),e.classList.add("dark"))}
    setColorMode(sessionStorage._theme||"auto");
  </script><link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=aecf457f" />
    <link rel="stylesheet" type="text/css" href="../_static/shibuya.css?v=5d7d6ab7" />
    <link media="print" rel="stylesheet" type="text/css" href="../_static/print.css?v=20ff2c19" />
    <link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
<style>
:root {
  --sy-f-text: "Inter", var(--sy-f-sys), var(--sy-f-cjk), sans-serif;
  --sy-f-heading: "Inter", var(--sy-f-sys), var(--sy-f-cjk), sans-serif;
}
</style>
    <meta property="og:type" content="website"/><meta property="og:title" content="Models"/>
<meta name="twitter:card" content="summary"/></head>
<body><div class="document"><div class="sy-head">
  <div class="sy-head-blur"></div>
  <div class="sy-head-inner sy-container mx-auto">
    <a class="sy-head-brand mr-4" href="../index.html">
      
      
      <strong>mango</strong>
    </a>
    <div class="sy-head-links" id="NavLinks">
      <nav class="sy-head-nav"></nav>
      <div class="sy-head-extra flex items-center print:hidden">
<form class="searchbox flex items-center" action="../search.html" method="get">
  <input type="text" name="q" placeholder="Search" />
  <kbd>/</kbd>
</form>
        
        
        <div class="sy-head-social flex items-center">
        </div>
      </div>
    </div>
    <div class="sy-head-actions flex items-center shrink-0 print:hidden">
      <button class="js-theme theme-switch flex items-center"
        data-aria-auto="Switch to light color mode"
        data-aria-light="Switch to dark color mode"
        data-aria-dark="Switch to auto color mode">
        <i class="i-icon theme-icon"></i>
      </button>
      <button class="md:hidden flex items-center js-menu" aria-label="Menu" type="button" aria-controls="NavLinks" aria-expanded="false">
        <div class="hamburger">
          <span class="hamburger_1"></span>
          <span class="hamburger_2 -translate-x-2"></span>
          <span class="hamburger_3 -translate-x-1"></span>
        </div>
      </button>
    </div>
  </div>
</div>
<div class="sy-page sy-container flex mx-auto">
  <aside id="lside" class="sy-lside md:w-72 md:shrink-0 print:hidden">
    <div class="sy-lside-inner md:sticky">
      <div class="sy-scrollbar p-6">
        <div class="globaltoc" data-expand-depth="0"><p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Mango</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genetic/index.html">Genetic Algorithms (GA)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../genetic/introduction.html">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/individual.html">Individual</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/population.html">Population</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/selection.html">Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/crossover.html">Crossover</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/mutation.html">Mutation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/replacement.html">Replacement</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/stop.html">Stop conditions</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/config.html">Configuration</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genetic/usage.html">Usage</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Code documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="benchmark.html">Benchmarks</a></li>
<li class="toctree-l2"><a class="reference internal" href="clients.html">Clients</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Models</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../bib.html">Bibliography</a></li>
</ul>

        </div>
      </div>
      
    </div>
  </aside>
  <div class="lside-overlay js-menu" role="button" aria-label="Close left sidebar" aria-controls="lside" aria-expanded="false"></div>
  <aside id="rside" class="sy-rside pb-3 w-64 shrink-0 order-last">
    <button class="rside-close js-menu xl:hidden" aria-label="Close Table of Contents" type="button" aria-controls="rside" aria-expanded="false">
      <i class="i-icon close"></i>
    </button>
    <div class="sy-scrollbar sy-rside-inner px-6 xl:top-16 xl:sticky xl:pl-0 pt-6 pb-4">
  <div class="localtoc">
    <h3>On this page</h3>
    <ul>
<li><a class="reference internal" href="#genetic">Genetic</a><ul>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual"><code class="docutils literal notranslate"><span class="pre">Individual</span></code></a><ul>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.genes"><code class="docutils literal notranslate"><span class="pre">Individual.genes</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.idx"><code class="docutils literal notranslate"><span class="pre">Individual.idx</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.config"><code class="docutils literal notranslate"><span class="pre">Individual.config</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.parents"><code class="docutils literal notranslate"><span class="pre">Individual.parents</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.encoding"><code class="docutils literal notranslate"><span class="pre">Individual.encoding</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.fitness"><code class="docutils literal notranslate"><span class="pre">Individual.fitness</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.create_random_individual"><code class="docutils literal notranslate"><span class="pre">Individual.create_random_individual()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.create_random_genes"><code class="docutils literal notranslate"><span class="pre">Individual.create_random_genes()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.mutate"><code class="docutils literal notranslate"><span class="pre">Individual.mutate()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual._mutate_binary"><code class="docutils literal notranslate"><span class="pre">Individual._mutate_binary()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual._mutate_integer"><code class="docutils literal notranslate"><span class="pre">Individual._mutate_integer()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual._mutate_real"><code class="docutils literal notranslate"><span class="pre">Individual._mutate_real()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.dominates"><code class="docutils literal notranslate"><span class="pre">Individual.dominates()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.copy"><code class="docutils literal notranslate"><span class="pre">Individual.copy()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.__hash__"><code class="docutils literal notranslate"><span class="pre">Individual.__hash__()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.individual.Individual.__eq__"><code class="docutils literal notranslate"><span class="pre">Individual.__eq__()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population"><code class="docutils literal notranslate"><span class="pre">Population</span></code></a><ul>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.config"><code class="docutils literal notranslate"><span class="pre">Population.config</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.evaluator"><code class="docutils literal notranslate"><span class="pre">Population.evaluator</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.run"><code class="docutils literal notranslate"><span class="pre">Population.run()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.continue_running"><code class="docutils literal notranslate"><span class="pre">Population.continue_running()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.init_population"><code class="docutils literal notranslate"><span class="pre">Population.init_population()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.selection"><code class="docutils literal notranslate"><span class="pre">Population.selection()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.crossover"><code class="docutils literal notranslate"><span class="pre">Population.crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.mutate"><code class="docutils literal notranslate"><span class="pre">Population.mutate()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.update_population"><code class="docutils literal notranslate"><span class="pre">Population.update_population()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.replace"><code class="docutils literal notranslate"><span class="pre">Population.replace()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.update_best"><code class="docutils literal notranslate"><span class="pre">Population.update_best()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.stop"><code class="docutils literal notranslate"><span class="pre">Population.stop()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._random_selection"><code class="docutils literal notranslate"><span class="pre">Population._random_selection()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._elitism_selection"><code class="docutils literal notranslate"><span class="pre">Population._elitism_selection()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._rank_selection"><code class="docutils literal notranslate"><span class="pre">Population._rank_selection()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._order_selection"><code class="docutils literal notranslate"><span class="pre">Population._order_selection()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._roulette_selection"><code class="docutils literal notranslate"><span class="pre">Population._roulette_selection()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._tournament_selection"><code class="docutils literal notranslate"><span class="pre">Population._tournament_selection()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._boltzmann_selection"><code class="docutils literal notranslate"><span class="pre">Population._boltzmann_selection()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._select_parents"><code class="docutils literal notranslate"><span class="pre">Population._select_parents()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._one_split_crossover"><code class="docutils literal notranslate"><span class="pre">Population._one_split_crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._two_split_crossover"><code class="docutils literal notranslate"><span class="pre">Population._two_split_crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._mask_crossover"><code class="docutils literal notranslate"><span class="pre">Population._mask_crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._linear_crossover"><code class="docutils literal notranslate"><span class="pre">Population._linear_crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._flat_crossover"><code class="docutils literal notranslate"><span class="pre">Population._flat_crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._blend_crossover"><code class="docutils literal notranslate"><span class="pre">Population._blend_crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._gaussian_crossover"><code class="docutils literal notranslate"><span class="pre">Population._gaussian_crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._morphology_crossover"><code class="docutils literal notranslate"><span class="pre">Population._morphology_crossover()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._add_offspring"><code class="docutils literal notranslate"><span class="pre">Population._add_offspring()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._base_mutation"><code class="docutils literal notranslate"><span class="pre">Population._base_mutation()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._adaptative_mutation"><code class="docutils literal notranslate"><span class="pre">Population._adaptative_mutation()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._calculate_coefficient_variation"><code class="docutils literal notranslate"><span class="pre">Population._calculate_coefficient_variation()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._update_fitness"><code class="docutils literal notranslate"><span class="pre">Population._update_fitness()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._random_replacement"><code class="docutils literal notranslate"><span class="pre">Population._random_replacement()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._offspring_replacement"><code class="docutils literal notranslate"><span class="pre">Population._offspring_replacement()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._elitist_replacement"><code class="docutils literal notranslate"><span class="pre">Population._elitist_replacement()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population._elitist_stochastic_replacement"><code class="docutils literal notranslate"><span class="pre">Population._elitist_stochastic_replacement()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.population.Population.extract_fitness"><code class="docutils literal notranslate"><span class="pre">Population.extract_fitness()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#mango.models.genetic.config.GeneticBaseConfig"><code class="docutils literal notranslate"><span class="pre">GeneticBaseConfig</span></code></a><ul>
<li><a class="reference internal" href="#mango.models.genetic.config.GeneticBaseConfig.__params"><code class="docutils literal notranslate"><span class="pre">GeneticBaseConfig.__params</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.config.GeneticBaseConfig.modify_value"><code class="docutils literal notranslate"><span class="pre">GeneticBaseConfig.modify_value()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#mango.models.genetic.problem.Problem"><code class="docutils literal notranslate"><span class="pre">Problem</span></code></a><ul>
<li><a class="reference internal" href="#mango.models.genetic.problem.Problem.calculate_fitness"><code class="docutils literal notranslate"><span class="pre">Problem.calculate_fitness()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.genetic.problem.Problem._abc_impl"><code class="docutils literal notranslate"><span class="pre">Problem._abc_impl</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#mango.models.genetic.shared.exceptions.GeneticDiversity"><code class="docutils literal notranslate"><span class="pre">GeneticDiversity</span></code></a><ul>
<li><a class="reference internal" href="#mango.models.genetic.shared.exceptions.GeneticDiversity.with_traceback"><code class="docutils literal notranslate"><span class="pre">GeneticDiversity.with_traceback()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#mango.models.genetic.shared.exceptions.ConfigurationError"><code class="docutils literal notranslate"><span class="pre">ConfigurationError</span></code></a><ul>
<li><a class="reference internal" href="#mango.models.genetic.shared.exceptions.ConfigurationError.with_traceback"><code class="docutils literal notranslate"><span class="pre">ConfigurationError.with_traceback()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#neural-networks-components">Neural Networks components</a><ul>
<li><a class="reference internal" href="#module-mango.models.activations">Activations</a><ul>
<li><a class="reference internal" href="#mango.models.activations.sigmoid"><code class="docutils literal notranslate"><span class="pre">sigmoid()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.activations.tanh"><code class="docutils literal notranslate"><span class="pre">tanh()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-mango.models.neural_networks">Neural networks</a><ul>
<li><a class="reference internal" href="#mango.models.neural_networks.calculate_network_output"><code class="docutils literal notranslate"><span class="pre">calculate_network_output()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#optimization">Optimization</a><ul>
<li><a class="reference internal" href="#module-mango.models.pyomo">Pyomo</a><ul>
<li><a class="reference internal" href="#mango.models.pyomo.is_feasible"><code class="docutils literal notranslate"><span class="pre">is_feasible()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.safe_value"><code class="docutils literal notranslate"><span class="pre">safe_value()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.var_to_sdict"><code class="docutils literal notranslate"><span class="pre">var_to_sdict()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.var_to_table"><code class="docutils literal notranslate"><span class="pre">var_to_table()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.get_status"><code class="docutils literal notranslate"><span class="pre">get_status()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.get_gap"><code class="docutils literal notranslate"><span class="pre">get_gap()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.write_cbc_warmstart_file"><code class="docutils literal notranslate"><span class="pre">write_cbc_warmstart_file()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.variables_to_excel"><code class="docutils literal notranslate"><span class="pre">variables_to_excel()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.variables_to_json"><code class="docutils literal notranslate"><span class="pre">variables_to_json()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.instance_from_excel"><code class="docutils literal notranslate"><span class="pre">instance_from_excel()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.instance_from_json"><code class="docutils literal notranslate"><span class="pre">instance_from_json()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.instance_from_dict"><code class="docutils literal notranslate"><span class="pre">instance_from_dict()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.load_variable"><code class="docutils literal notranslate"><span class="pre">load_variable()</span></code></a></li>
<li><a class="reference internal" href="#mango.models.pyomo.model_data_to_excel"><code class="docutils literal notranslate"><span class="pre">model_data_to_excel()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div><div id="ethical-ad-placement" data-ea-publisher="readthedocs"></div>
    </div>
  </aside>
  <div class="rside-overlay js-menu" role="button" aria-label="Close Table of Contents" aria-controls="rside" aria-expanded="false"></div>
  <main class="sy-main w-full max-sm:max-w-full print:pt-6">
    <div class="sy-breadcrumbs" role="navigation">
  <div class="sy-breadcrumbs-inner flex items-center">
    <div class="md:hidden mr-3">
      <button class="js-menu" aria-label="Menu" type="button" aria-controls="lside" aria-expanded="false">
        <i class="i-icon menu"></i>
      </button>
    </div>
    <ul class="flex-1">
      
      
      <li><a href="../index.html">mango</a><span>/</span></li>
      
      <li><a href="index.html">Code documentation</a><span>/</span></li>
      
      <li><strong>Models</strong></li>
      
      
    </ul>
    <div class="xl:hidden ml-1">
      <button class="js-menu" aria-label="Show table of contents" type="button" aria-controls="rside"
        aria-expanded="false">
        <i class="i-icon outdent"></i>
      </button>
    </div>
  </div>
</div>
    <div class="flex flex-col break-words justify-between">
      <div class="min-w-0 max-w-6xl px-6 pb-6 pt-8 xl:px-12">
        <article class="yue" role="main">
          <section id="models">
<h1>Models<a class="headerlink" href="#models" title="Permalink to this heading">¶</a></h1>
<section id="genetic">
<h2>Genetic<a class="headerlink" href="#genetic" title="Permalink to this heading">¶</a></h2>
<p>Here we have the code documentation for the main classes of the genetic algorithm: individual, population, configuration and problem.</p>
<dl class="py class">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">mango.models.genetic.individual.</span></span><span class="sig-name descname"><span class="pre">Individual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.individual.Individual" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py property">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ndarray</span></em><a class="headerlink" href="#mango.models.genetic.individual.Individual.genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Property that stores the <code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code> of genes in the individual</p>
<p>If max_bound or min_bound are set, the genes values are clipped to the bounds.
After setting the value for the genes the hash is updated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code>) – the new value for the genes (used on setting up the property)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the genes value</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.idx">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">idx</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#mango.models.genetic.individual.Individual.idx" title="Permalink to this definition">¶</a></dt>
<dd><p>Property that stores the internal idx of the Individual. This value is just to help
distinguish between individuals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (<em>int</em>) – the new value for the idx (used on setting up the property)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the idx value</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.config">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#mango.models.genetic.config.GeneticBaseConfig" title="mango.models.genetic.config.config.GeneticBaseConfig"><span class="pre">GeneticBaseConfig</span></a></em><a class="headerlink" href="#mango.models.genetic.individual.Individual.config" title="Permalink to this definition">¶</a></dt>
<dd><p>Property that stores the config object for the individual</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (<a class="reference internal" href="#mango.models.genetic.config.GeneticBaseConfig" title="mango.models.genetic.config.GeneticBaseConfig"><code class="xref py py-class docutils literal notranslate"><span class="pre">mango.models.genetic.config.GeneticBaseConfig</span></code></a>) – the new value for the config (used on setting up the property)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the config value</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.parents">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parents</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">tuple</span></em><a class="headerlink" href="#mango.models.genetic.individual.Individual.parents" title="Permalink to this definition">¶</a></dt>
<dd><p>Property that stores the parents of the individual and their idx.</p>
<p>During the setter the value for attribute parents_idx
is updated as well as a tuple of the idx of all the parents</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (<em>tuple</em>) – the new value for the parents (used on setting up the property)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the parents value</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.encoding">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">encoding</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#mango.models.genetic.individual.Individual.encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Property that stores the encoding type for the individual</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (<em>str</em>) – the new value for the encoding (used on setting up the property)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the encoding value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.fitness">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fitness</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">float</span></em><a class="headerlink" href="#mango.models.genetic.individual.Individual.fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Property that stores the fitness value for the individual</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (<em>float</em>) – the new value for the fitness (used on setting up the property)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the fitness value</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.create_random_individual">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">create_random_individual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#mango.models.genetic.config.GeneticBaseConfig" title="mango.models.genetic.config.config.GeneticBaseConfig"><span class="pre">GeneticBaseConfig</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#mango.models.genetic.individual.Individual" title="mango.models.genetic.individual.base_individual.Individual"><span class="pre">Individual</span></a></span></span><a class="headerlink" href="#mango.models.genetic.individual.Individual.create_random_individual" title="Permalink to this definition">¶</a></dt>
<dd><p>Class method that creates a new individual with random genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>idx</strong> (<em>int</em>) – the idx of the new individual</p></li>
<li><p><strong>config</strong> (<a class="reference internal" href="#mango.models.genetic.config.GeneticBaseConfig" title="mango.models.genetic.config.GeneticBaseConfig"><code class="xref py py-class docutils literal notranslate"><span class="pre">mango.models.genetic.config.GeneticBaseConfig</span></code></a>) – the config object for the individual</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the new individual</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.create_random_genes">
<span class="sig-name descname"><span class="pre">create_random_genes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.individual.Individual.create_random_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to create random genes for an individual based on the encoding type.</p>
<p>The genes are stored in the <a class="reference internal" href="#mango.models.genetic.individual.Individual.genes" title="mango.models.genetic.individual.Individual.genes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">genes</span></code></a> attribute</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.mutate">
<span class="sig-name descname"><span class="pre">mutate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation_prob</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.individual.Individual.mutate" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to mutate the individual based on the encoding type.</p>
<p>The mutation probability is used to determine if the individual will be mutated or not.
An individual can have more than one gene that gets mutated. The</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mutation_prob</strong> (<em>float</em>) – the probability of mutation</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual._mutate_binary">
<span class="sig-name descname"><span class="pre">_mutate_binary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.individual.Individual._mutate_binary" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutation method for the binary values encoding.</p>
<p>This mutation is a simple bit flip on the calculated gene position</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual._mutate_integer">
<span class="sig-name descname"><span class="pre">_mutate_integer</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.individual.Individual._mutate_integer" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutation method for the integer values encoding.</p>
<p>This mutation is a simple random value between the min and max bounds (randint)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual._mutate_real">
<span class="sig-name descname"><span class="pre">_mutate_real</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.individual.Individual._mutate_real" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutation method for the real values encoding</p>
<p>This mutation is a simple random value between the min and max bounds</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.dominates">
<span class="sig-name descname"><span class="pre">dominates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.individual.Individual.dominates" title="Permalink to this definition">¶</a></dt>
<dd><p>This method should implement the logic to check if one individual dominates another one</p>
<p>The domination concept is mainly used in multiobjective optimization problems.
A solution dominates another when the first one is equal or better in all objectives and one objective
is at least better than the other solution</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#mango.models.genetic.individual.Individual" title="mango.models.genetic.individual.base_individual.Individual"><span class="pre">Individual</span></a></span></span><a class="headerlink" href="#mango.models.genetic.individual.Individual.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to create a copy of the individual</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.individual.Individual.__hash__" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the hash of the individual and store it in the <code class="xref py py-attr docutils literal notranslate"><span class="pre">_hash</span></code> attribute</p>
<p>The hash is calculated based on the byte representation of the genes of the individual so that only individuals
with the same genome have the hash</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.individual.Individual.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#mango.models.genetic.individual.Individual.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to compare two individuals</p>
<p>In order to make this comparison as fast as possible it used the pre-generated hash to compare the individuals.
So the comparison is made on the genotype instead of the phenotype, as two individuals with different
genotypes can have the same phenotype, so they should be considered as not equal
for crossover and genetic diversity purposes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> – the other individual to compare</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the individuals are equal, False otherwise</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">mango.models.genetic.population.</span></span><span class="sig-name descname"><span class="pre">Population</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">config</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">evaluator</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py property">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.config">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">config</span></span><a class="headerlink" href="#mango.models.genetic.population.Population.config" title="Permalink to this definition">¶</a></dt>
<dd><p>Property to store the configuration that the genetic algorithm has to use.</p>
<dl class="field-list simple">
<dt class="field-odd">Getter<span class="colon">:</span></dt>
<dd class="field-odd"><p>Returns the configuration object.</p>
</dd>
<dt class="field-even">Setter<span class="colon">:</span></dt>
<dd class="field-even"><p>Sets the configuration object.</p>
</dd>
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (<a class="reference internal" href="#mango.models.genetic.config.GeneticBaseConfig" title="mango.models.genetic.config.GeneticBaseConfig"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneticBaseConfig</span></code></a>) – The configuration object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.evaluator">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evaluator</span></span><a class="headerlink" href="#mango.models.genetic.population.Population.evaluator" title="Permalink to this definition">¶</a></dt>
<dd><p>Property to store the function or class that evaluates the fitness of an individual.</p>
<dl class="field-list simple">
<dt class="field-odd">Getter<span class="colon">:</span></dt>
<dd class="field-odd"><p>Returns the evaluator function.</p>
</dd>
<dt class="field-even">Setter<span class="colon">:</span></dt>
<dd class="field-even"><p>Sets the evaluator function.</p>
</dd>
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (Union[callable, <a class="reference internal" href="#mango.models.genetic.problem.Problem" title="mango.models.genetic.problem.Problem"><code class="xref py py-class docutils literal notranslate"><span class="pre">Problem</span></code></a>]) – The evaluator function or class.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to run the evolution process on the population.</p>
<p>This method does not have any input argument as all the information should be passed through the config class.
This method does not return anything, but it can be stopped due to the exceptions that can be raised on
other parts of the algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#mango.models.genetic.shared.exceptions.GeneticDiversity" title="mango.models.genetic.shared.exceptions.GeneticDiversity"><strong>GeneticDiversity</strong></a> – When the genetic diversity is too low this exception gets raised. it can be raised
during the parent selection process or during the stop phase of the algorithm.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.continue_running">
<span class="sig-name descname"><span class="pre">continue_running</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">generations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.continue_running" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to rerun the Genetic Algorithm from the point it stopped for more generations.</p>
<p>This method only works if the algorithm has previously stopped due to reaching the initial generation limit,
if it is due to a <a class="reference internal" href="#mango.models.genetic.shared.exceptions.GeneticDiversity" title="mango.models.genetic.shared.exceptions.GeneticDiversity"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneticDiversity</span></code></a> exception,
it will probably raise said exception again.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>generations</strong> – Number of generations to run the Genetic Algorithm.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#mango.models.genetic.shared.exceptions.GeneticDiversity" title="mango.models.genetic.shared.exceptions.GeneticDiversity"><strong>GeneticDiversity</strong></a> – When the genetic diversity is too low this exception gets raised. it can be raised
during the parent selection process or during the stop phase of the algorithm.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.init_population">
<span class="sig-name descname"><span class="pre">init_population</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.init_population" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to initialize the population in the Genetic Algorithm</p>
<p>The initialization of the population is done by creating a list of random individuals with size
equal to the population_size parameter.
The creation of the individuals is controlled by the create_random_individual class method of the
<a class="reference internal" href="#mango.models.genetic.individual.Individual" title="mango.models.genetic.individual.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code></a> class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.selection">
<span class="sig-name descname"><span class="pre">selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to run the selection phase of the Genetic Algorithm.
The selection process actually just creates the list of probabilities of a given individual
to be chosen on the crossover phase. The values of said probabilities depend on the selection method chosen..</p>
<p>Currently, there is seven selection methods implemented:
- Random (<a class="reference internal" href="#mango.models.genetic.population.Population._random_selection" title="mango.models.genetic.population.Population._random_selection"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_random_selection()</span></code></a>).
- Elitism (<a class="reference internal" href="#mango.models.genetic.population.Population._elitism_selection" title="mango.models.genetic.population.Population._elitism_selection"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_elitism_selection()</span></code></a>).
- Rank (<a class="reference internal" href="#mango.models.genetic.population.Population._rank_selection" title="mango.models.genetic.population.Population._rank_selection"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_rank_selection()</span></code></a>).
- Order (<a class="reference internal" href="#mango.models.genetic.population.Population._order_selection" title="mango.models.genetic.population.Population._order_selection"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_order_selection()</span></code></a>).
- Roulette (<a class="reference internal" href="#mango.models.genetic.population.Population._roulette_selection" title="mango.models.genetic.population.Population._roulette_selection"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_roulette_selection()</span></code></a>).
- Tournament (<a class="reference internal" href="#mango.models.genetic.population.Population._tournament_selection" title="mango.models.genetic.population.Population._tournament_selection"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_tournament_selection()</span></code></a>)
- Boltzmann (<a class="reference internal" href="#mango.models.genetic.population.Population._boltzmann_selection" title="mango.models.genetic.population.Population._boltzmann_selection"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_boltzmann_selection()</span></code></a>, not implemented yet).</p>
<p>The details of how each of these methods work can be found on the <a class="reference internal" href="../genetic/selection.html#selection-label"><span class="std std-ref">Selection</span></a> documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.crossover">
<span class="sig-name descname"><span class="pre">crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to run the crossover phase of the Genetic Algorithm</p>
<p>Currently, there is eight crossover method implemented:
- One-split (<a class="reference internal" href="#mango.models.genetic.population.Population._one_split_crossover" title="mango.models.genetic.population.Population._one_split_crossover"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_one_split_crossover()</span></code></a>).
- Two-split (<a class="reference internal" href="#mango.models.genetic.population.Population._two_split_crossover" title="mango.models.genetic.population.Population._two_split_crossover"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_two_split_crossover()</span></code></a>).
- Mask (<a class="reference internal" href="#mango.models.genetic.population.Population._mask_crossover" title="mango.models.genetic.population.Population._mask_crossover"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_mask_crossover()</span></code></a>).
- Linear (<a class="reference internal" href="#mango.models.genetic.population.Population._linear_crossover" title="mango.models.genetic.population.Population._linear_crossover"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_linear_crossover()</span></code></a>).
- Flat (<a class="reference internal" href="#mango.models.genetic.population.Population._flat_crossover" title="mango.models.genetic.population.Population._flat_crossover"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_flat_crossover()</span></code></a>).
- Blend (<a class="reference internal" href="#mango.models.genetic.population.Population._blend_crossover" title="mango.models.genetic.population.Population._blend_crossover"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_blend_crossover()</span></code></a>).
- Gaussian (<a class="reference internal" href="#mango.models.genetic.population.Population._gaussian_crossover" title="mango.models.genetic.population.Population._gaussian_crossover"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_gaussian_crossover()</span></code></a>).
- Morphology (<a class="reference internal" href="#mango.models.genetic.population.Population._morphology_crossover" title="mango.models.genetic.population.Population._morphology_crossover"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_morphology_crossover()</span></code></a>, not implemented yet).</p>
<p>The details of how each of these cross over works can be found on the <a class="reference internal" href="../genetic/crossover.html#crossover-label"><span class="std std-ref">Crossover</span></a> documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.mutate">
<span class="sig-name descname"><span class="pre">mutate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.mutate" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to run the mutation phase of the Genetic Algorithm.</p>
<p>The actual implementation of the mutation is done in the individual as it is heavily dependent on the encoding.
The only change is when using adaptative mutation that the mutation rate has to be updated before calling
the individual.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.update_population">
<span class="sig-name descname"><span class="pre">update_population</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.update_population" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to make sure that all individuals in the population have a fitness value.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.replace" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to run the replacement phase of the Genetic Algorithm.</p>
<p>Currently, there is one replacement method implemented:
- Random (<a class="reference internal" href="#mango.models.genetic.population.Population._random_replacement" title="mango.models.genetic.population.Population._random_replacement"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_random_replacement()</span></code></a>).
- Offspring (<a class="reference internal" href="#mango.models.genetic.population.Population._offspring_replacement" title="mango.models.genetic.population.Population._offspring_replacement"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_offspring_replacement()</span></code></a>).
- Elitist (<a class="reference internal" href="#mango.models.genetic.population.Population._elitist_replacement" title="mango.models.genetic.population.Population._elitist_replacement"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_elitist_replacement()</span></code></a>).
- Elistist stochastic (<a class="reference internal" href="#mango.models.genetic.population.Population._elitist_stochastic_replacement" title="mango.models.genetic.population.Population._elitist_stochastic_replacement"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_elitist_stochastic_replacement()</span></code></a>).</p>
<p>The details of how each of these replacement methods work can be found on the <a class="reference internal" href="../genetic/replacement.html#replacement-label"><span class="std std-ref">Replacement</span></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.update_best">
<span class="sig-name descname"><span class="pre">update_best</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.update_best" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to update the best adapted individual on the population.</p>
<p>The best individual gets updated based on the optimization objective of the problem.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.stop">
<span class="sig-name descname"><span class="pre">stop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.stop" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to implement stop conditions based on information about the population or the genetic diversity.</p>
<p>Currently, the only check done is that the genetic diversity is not too low.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._random_selection">
<span class="sig-name descname"><span class="pre">_random_selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._random_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform random selection.</p>
<p>In this case the probability to select any individual is equal.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._elitism_selection">
<span class="sig-name descname"><span class="pre">_elitism_selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._elitism_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform elitism selection.</p>
<p>Based on the config parameter rank_size (k), the k-best individuals are selected for crossover.
That means that k individuals will have the same selection probability and the rest will have zero.</p>
<p>This selection method could create a diversity problem on the long run if k is small compared
to the population size.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._rank_selection">
<span class="sig-name descname"><span class="pre">_rank_selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._rank_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform rank selection.</p>
<p>This method is based on the rank of the individuals on the population.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._order_selection">
<span class="sig-name descname"><span class="pre">_order_selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._order_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform order selection</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._roulette_selection">
<span class="sig-name descname"><span class="pre">_roulette_selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._roulette_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform roulette selection.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._tournament_selection">
<span class="sig-name descname"><span class="pre">_tournament_selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._tournament_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform tournament selection</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._boltzmann_selection">
<span class="sig-name descname"><span class="pre">_boltzmann_selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._boltzmann_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Not implemented yet</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._select_parents">
<span class="sig-name descname"><span class="pre">_select_parents</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span></span></span><a class="headerlink" href="#mango.models.genetic.population.Population._select_parents" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to select n parents from the population based on the selection probabilities
established on the selection phase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em>) – number of parents to select from the population. It defaults to two, but it has the capability
to be a different number.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a tuple with the selected parents</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._one_split_crossover">
<span class="sig-name descname"><span class="pre">_one_split_crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._one_split_crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform one-split crossover.</p>
<p>This method creates two children individuals from two parents.
It selects a random split point and creates two children by combining the genes of the parents.</p>
<p>More details of this crossover can be found on the <a class="reference internal" href="../genetic/crossover.html#one-split-label"><span class="std std-ref">One-split crossover</span></a> section of the documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._two_split_crossover">
<span class="sig-name descname"><span class="pre">_two_split_crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._two_split_crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform two-split crossover.</p>
<p>This method creates two children individuals from two parents.
It selects two random split points and creates two children by combining the genes of the parents.</p>
<p>More details of this crossover can be found on the <a class="reference internal" href="../genetic/crossover.html#two-split-label"><span class="std std-ref">Two-split crossover</span></a> section of the documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._mask_crossover">
<span class="sig-name descname"><span class="pre">_mask_crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._mask_crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform mask crossover.</p>
<p>This method creates two children individuals from two parents.
It selects a random mask and creates two children by combining the genes of the parents.
The first children is created by taking the genes of the first parent where the mask is 1 and the genes of
the second parent where the mask is 0. The second children is created by taking the genes of the first parent
where the mask is 0 and the genes of the second parent where the mask is 1.</p>
<p>More details of this crossover can be found on the <a class="reference internal" href="../genetic/crossover.html#mask-label"><span class="std std-ref">Mask crossover</span></a> section of the documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._linear_crossover">
<span class="sig-name descname"><span class="pre">_linear_crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._linear_crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform the linear crossover proposed by Wright (1991).</p>
<p>This method creates three child individuals. The first child is calculated as the midpoint between both parents.
The other two lie on the line defined by the parents are calculated as:</p>
<ul class="simple">
<li><p>child2 = 1.5 * p1 - 0.5 * p2</p></li>
<li><p>child3 = -0.5 * p1 + 1.5 * p2</p></li>
</ul>
<p>More details of this crossover can be found on the <a class="reference internal" href="../genetic/crossover.html#linear-label"><span class="std std-ref">Linear crossover</span></a> section of the documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._flat_crossover">
<span class="sig-name descname"><span class="pre">_flat_crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._flat_crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>This method implements Radcliffe’s flat crossover (1990).</p>
<p>It creates two child individuals with each gene created randomly inside the interval defined by the parents.</p>
<p>More details of this crossover can be found on the <a class="reference internal" href="../genetic/crossover.html#flat-label"><span class="std std-ref">Flat crossover</span></a> section of the documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._blend_crossover">
<span class="sig-name descname"><span class="pre">_blend_crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._blend_crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>This method implements the Blend crossover proposed by Eshelman and Schaffer (1993).</p>
<p>The alpha parameter defined for the crossover is defined as blend_expansion parameter.
With a value of 0 it is equal to a Radcliffe’s flat crossover while a value of 0.5 is
equal to Wright’s linear crossover intervals (but the values are randomly selected).</p>
<p>The main difference with Wrights linear crossover is that only two childs are created instead of two.</p>
<p>More details of this crossover can be found on the <a class="reference internal" href="../genetic/crossover.html#blend-label"><span class="std std-ref">Blend crossover</span></a> section of the documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._gaussian_crossover">
<span class="sig-name descname"><span class="pre">_gaussian_crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._gaussian_crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to perform the Gaussian crossover, also known as Unimodal Normal Distribution Crossover
or UNDX, proposed by Ono (2003).</p>
<p>This method is only implemented for real encoding.</p>
<p>More details of this crossover can be found on the <a class="reference internal" href="../genetic/crossover.html#gaussian-label"><span class="std std-ref">Gaussian crossover</span></a> section of the documentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._morphology_crossover">
<span class="sig-name descname"><span class="pre">_morphology_crossover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._morphology_crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is not yet implemented</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._add_offspring">
<span class="sig-name descname"><span class="pre">_add_offspring</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">parents</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._add_offspring" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to add an offspring to the offspring array</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genes</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.array</span></code>) – Genes of the offspring</p></li>
<li><p><strong>parents</strong> (<em>tuple</em>) – Parents of the offspring</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._base_mutation">
<span class="sig-name descname"><span class="pre">_base_mutation</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._base_mutation" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the based mutation applied for the base mutation rate or
the ones calculated based on gene length or population size.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._adaptative_mutation">
<span class="sig-name descname"><span class="pre">_adaptative_mutation</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._adaptative_mutation" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to implement adaptative mutation operator.</p>
<p>For the update of the mutation rate we use the coefficient of variation of the fitness of the population.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._calculate_coefficient_variation">
<span class="sig-name descname"><span class="pre">_calculate_coefficient_variation</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._calculate_coefficient_variation" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the coefficient of variation of the fitness of the population.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._update_fitness">
<span class="sig-name descname"><span class="pre">_update_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">population</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._update_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Internal method to update the fitness of a gioven population or subpopulation of the GA.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._random_replacement">
<span class="sig-name descname"><span class="pre">_random_replacement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._random_replacement" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to implement a random replacement operator.</p>
<p>In this case the offspring is added to the population and then a random selection is done to select
the individuals that will pass to the next generation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._offspring_replacement">
<span class="sig-name descname"><span class="pre">_offspring_replacement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._offspring_replacement" title="Permalink to this definition">¶</a></dt>
<dd><p>The population gets completely substituted by the offspring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._elitist_replacement">
<span class="sig-name descname"><span class="pre">_elitist_replacement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._elitist_replacement" title="Permalink to this definition">¶</a></dt>
<dd><p>The best adapted individuals on the population are the ones that pass to the next generation</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population._elitist_stochastic_replacement">
<span class="sig-name descname"><span class="pre">_elitist_stochastic_replacement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population._elitist_stochastic_replacement" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to implement an elitist stochastic replacement operator.</p>
<p>In this case each individual has a replacement probability based on their fitness.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.population.Population.extract_fitness">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">extract_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.population.Population.extract_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper static method to extract the fitness of an individual.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="mango.models.genetic.config.GeneticBaseConfig">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">mango.models.genetic.config.</span></span><span class="sig-name descname"><span class="pre">GeneticBaseConfig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.config.GeneticBaseConfig" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">BaseConfig</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="mango.models.genetic.config.GeneticBaseConfig.__params">
<span class="sig-name descname"><span class="pre">__params</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'crossover':</span> <span class="pre">[ConfigParameter('offspring_size',</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">100),</span> <span class="pre">ConfigParameter('blend_expansion',</span> <span class="pre">&lt;class</span> <span class="pre">'float'&gt;,</span> <span class="pre">0.1)],</span> <span class="pre">'individual':</span> <span class="pre">[ConfigParameter('encoding',</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;),</span> <span class="pre">ConfigParameter('gene_length',</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">0),</span> <span class="pre">ConfigParameter('gene_min_value',</span> <span class="pre">&lt;class</span> <span class="pre">'float'&gt;),</span> <span class="pre">ConfigParameter('gene_max_value',</span> <span class="pre">&lt;class</span> <span class="pre">'float'&gt;)],</span> <span class="pre">'main':</span> <span class="pre">[ConfigParameter('population_size',</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">100),</span> <span class="pre">ConfigParameter('max_generations',</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">500),</span> <span class="pre">ConfigParameter('optimization_objective',</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;),</span> <span class="pre">ConfigParameter('selection',</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;),</span> <span class="pre">ConfigParameter('crossover',</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;),</span> <span class="pre">ConfigParameter('replacement',</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;),</span> <span class="pre">ConfigParameter('mutation_control',</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;,</span> <span class="pre">'none'),</span> <span class="pre">ConfigParameter('mutation_base_rate',</span> <span class="pre">&lt;class</span> <span class="pre">'float'&gt;,</span> <span class="pre">0.1)],</span> <span class="pre">'mutation':</span> <span class="pre">[ConfigParameter('generation_adaptative',</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">10)],</span> <span class="pre">'selection':</span> <span class="pre">[ConfigParameter('elitism_size',</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">20),</span> <span class="pre">ConfigParameter('tournament_size',</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">2),</span> <span class="pre">ConfigParameter('rank_pressure',</span> <span class="pre">&lt;class</span> <span class="pre">'float'&gt;,</span> <span class="pre">2.0)]}</span></em><a class="headerlink" href="#mango.models.genetic.config.GeneticBaseConfig.__params" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.config.GeneticBaseConfig.modify_value">
<span class="sig-name descname"><span class="pre">modify_value</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.config.GeneticBaseConfig.modify_value" title="Permalink to this definition">¶</a></dt>
<dd><p>The modify value allows to modify the value of a parameter after it has been loaded</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>key</strong> – the name of the parameter we want to extract</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The value of the parameter</p>
</dd>
<dt class="field-odd">Doc-author<span class="colon">:</span></dt>
<dd class="field-odd"><p>baobab soluciones</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="mango.models.genetic.problem.Problem">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">mango.models.genetic.problem.</span></span><span class="sig-name descname"><span class="pre">Problem</span></span><a class="headerlink" href="#mango.models.genetic.problem.Problem" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">ABC</span></code></p>
<p>Metaclass to implement an abstract problem class for genetic algorithms.</p>
<p>The problem class is used to define the fitness function for the genetic algorithm.</p>
<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.problem.Problem.calculate_fitness">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">calculate_fitness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#mango.models.genetic.problem.Problem.calculate_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the fitness of a given solution.</p>
<p>This method has to be implemented on the subclasses</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>x</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.array</span></code>) – Solution to calculate the fitness of.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Fitness of the solution.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="mango.models.genetic.problem.Problem._abc_impl">
<span class="sig-name descname"><span class="pre">_abc_impl</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">&lt;_abc_data</span> <span class="pre">object&gt;</span></em><a class="headerlink" href="#mango.models.genetic.problem.Problem._abc_impl" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<p>There is also some other helper classes that are used by the genetic algorithm that are documented below.</p>
<dl class="py class">
<dt class="sig sig-object py" id="mango.models.genetic.shared.exceptions.GeneticDiversity">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">mango.models.genetic.shared.exceptions.</span></span><span class="sig-name descname"><span class="pre">GeneticDiversity</span></span><a class="headerlink" href="#mango.models.genetic.shared.exceptions.GeneticDiversity" title="Permalink to this definition">¶</a></dt>
<dd><p>Exception raised when the genetic diversity of the population is too low.</p>
<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.shared.exceptions.GeneticDiversity.with_traceback">
<span class="sig-name descname"><span class="pre">with_traceback</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.shared.exceptions.GeneticDiversity.with_traceback" title="Permalink to this definition">¶</a></dt>
<dd><p>Exception.with_traceback(tb) –
set self.__traceback__ to tb and return self.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="mango.models.genetic.shared.exceptions.ConfigurationError">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">mango.models.genetic.shared.exceptions.</span></span><span class="sig-name descname"><span class="pre">ConfigurationError</span></span><a class="headerlink" href="#mango.models.genetic.shared.exceptions.ConfigurationError" title="Permalink to this definition">¶</a></dt>
<dd><p>Exception raised when the configuration of the genetic algorithm is not valid.</p>
<dl class="py method">
<dt class="sig sig-object py" id="mango.models.genetic.shared.exceptions.ConfigurationError.with_traceback">
<span class="sig-name descname"><span class="pre">with_traceback</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.genetic.shared.exceptions.ConfigurationError.with_traceback" title="Permalink to this definition">¶</a></dt>
<dd><p>Exception.with_traceback(tb) –
set self.__traceback__ to tb and return self.</p>
</dd></dl>

</dd></dl>

</section>
<section id="neural-networks-components">
<h2>Neural Networks components<a class="headerlink" href="#neural-networks-components" title="Permalink to this heading">¶</a></h2>
<section id="module-mango.models.activations">
<span id="activations"></span><h3>Activations<a class="headerlink" href="#module-mango.models.activations" title="Permalink to this heading">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="mango.models.activations.sigmoid">
<span class="sig-prename descclassname"><span class="pre">mango.models.activations.</span></span><span class="sig-name descname"><span class="pre">sigmoid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">factor</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#mango.models.activations.sigmoid" title="Permalink to this definition">¶</a></dt>
<dd><p>Sigmoid funtion with a factor to control the slope</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>float</em>) – the input value</p></li>
<li><p><strong>factor</strong> (<em>float</em><em>, </em><em>optional</em>) – the factor to control the slope, defaults to 1</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the sigmoid value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.activations.tanh">
<span class="sig-prename descclassname"><span class="pre">mango.models.activations.</span></span><span class="sig-name descname"><span class="pre">tanh</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">factor</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#mango.models.activations.tanh" title="Permalink to this definition">¶</a></dt>
<dd><p>Tanh funtion with a factor to control the slope</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>float</em>) – the input value</p></li>
<li><p><strong>factor</strong> (<em>float</em><em>, </em><em>optional</em>) – the factor to control the slope, defaults to 1</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the tanh value</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-mango.models.neural_networks">
<span id="neural-networks"></span><h3>Neural networks<a class="headerlink" href="#module-mango.models.neural_networks" title="Permalink to this heading">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="mango.models.neural_networks.calculate_network_output">
<span class="sig-prename descclassname"><span class="pre">mango.models.neural_networks.</span></span><span class="sig-name descname"><span class="pre">calculate_network_output</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_nodes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hidden_nodes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_nodes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">array</span></span></span><a class="headerlink" href="#mango.models.neural_networks.calculate_network_output" title="Permalink to this definition">¶</a></dt>
<dd><p>The calculate_network_output function takes in a set of inputs, weights, and nodes.
It then calculates the output of the network by multiplying each input with its corresponding weight
and adding it to a bias term for that layer. The function then passes this value through an activation function
(sigmoid) and returns the result.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">np.array</span></code>) – the input data</p></li>
<li><p><strong>weights</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">np.array</span></code>) – the weights of each layer</p></li>
<li><p><strong>layers</strong> (<em>int</em>) – the number of layers of the network</p></li>
<li><p><strong>input_nodes</strong> (<em>int</em>) – the number of input nodes</p></li>
<li><p><strong>hidden_nodes</strong> (<em>list</em>) – the list with the number of nodes in each layer</p></li>
<li><p><strong>output_nodes</strong> (<em>int</em>) – the number of output nodes</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the result of the network for a given set of inputs</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">np.array</span></code></p>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="optimization">
<h2>Optimization<a class="headerlink" href="#optimization" title="Permalink to this heading">¶</a></h2>
<section id="module-mango.models.pyomo">
<span id="pyomo"></span><h3>Pyomo<a class="headerlink" href="#module-mango.models.pyomo" title="Permalink to this heading">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.is_feasible">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">is_feasible</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">status</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.is_feasible" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if the status is optimal or maxTimeLimit</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>status</strong> – a status (string or pyomo object)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the status is optimal or maxTimeLimit</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.safe_value">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">safe_value</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.safe_value" title="Permalink to this definition">¶</a></dt>
<dd><p>Safely apply pyomo value to a variable.</p>
<p>pyomo value generate an error if the variable has not been used.
This function will return 0 instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>x</strong> – a pyomo variable</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.var_to_sdict">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">var_to_sdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rounding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.var_to_sdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Transform a pyomo variable into a python dict</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variable</strong> – a pyomo variable.</p></li>
<li><p><strong>clean</strong> – if true, only return non-zero values.</p></li>
<li><p><strong>rounding</strong> – number of decimal to round.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a SuperDict containing the indices and values of the variable.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.var_to_table">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">var_to_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rounding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.var_to_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a table from a pyomo variable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variable</strong> – a pyomo variable.</p></li>
<li><p><strong>keys</strong> – the names of the keys of the variable.</p></li>
<li><p><strong>clean</strong> – if true, only return non-zero values.</p></li>
<li><p><strong>rounding</strong> – number of decimal to round.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a table [{Index_1:i1, Index_2:i2, Value:v}…]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.get_status">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">get_status</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">result</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.get_status" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the status of the solution from the result object</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>result</strong> – a pyomo result object</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the status</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.get_gap">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">get_gap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">result</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.get_gap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the relative gap of the solution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>result</strong> – a pyomo result object</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the gap</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.write_cbc_warmstart_file">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">write_cbc_warmstart_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">instance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">opt</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.write_cbc_warmstart_file" title="Permalink to this definition">¶</a></dt>
<dd><p>This function write a file to be passed to cbc solver as a warmstart file.
This function is necessary because of a bug of cbc that does not allow reading warmstart files on windows
with absolute path.
:param filename: path to the file
:param instance: model instance (created with create_instance)
:param opt: solver instance (created with solver factory)
:return:</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.variables_to_excel">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">variables_to_excel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_solution</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variables.xlsx'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rounding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.variables_to_excel" title="Permalink to this definition">¶</a></dt>
<dd><p>Save all the values of the variables in an Excel file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model_solution</strong> – the pyomo solution.</p></li>
<li><p><strong>path</strong> – path of the directory where to save the file</p></li>
<li><p><strong>clean</strong> – discard variables with values of 0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>nothing</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.variables_to_json">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">variables_to_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_solution</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variables.json'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.variables_to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Save all the values of the variables in an json file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model_solution</strong> – the pyomo solution.</p></li>
<li><p><strong>path</strong> – path of the directory where to save the file</p></li>
<li><p><strong>clean</strong> – discard variables with values of 0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>nothing</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.instance_from_excel">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">instance_from_excel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">instance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variables.xlsx'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.instance_from_excel" title="Permalink to this definition">¶</a></dt>
<dd><p>The function reads instance from excel.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>instance</strong> – model instance</p></li>
<li><p><strong>path</strong> – the path of the file to be loaded</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>model instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.instance_from_json">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">instance_from_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">instance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variables.json'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.instance_from_json" title="Permalink to this definition">¶</a></dt>
<dd><p>The function reads instance from json.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>instance</strong> – model instance</p></li>
<li><p><strong>path</strong> – the path of the file to be loaded</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>model instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.instance_from_dict">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">instance_from_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">instance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.instance_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>The function loads the instance based on data from dict with tables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>instance</strong> – model instance</p></li>
<li><p><strong>data</strong> – data for instance</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>model instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.load_variable">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">load_variable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">instance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.load_variable" title="Permalink to this definition">¶</a></dt>
<dd><p>The function loads the values to indicated variable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>instance</strong> – model instance</p></li>
<li><p><strong>var_name</strong> – name of the variable to which values will be assign</p></li>
<li><p><strong>value</strong> – values to be assigned to variable</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="mango.models.pyomo.model_data_to_excel">
<span class="sig-prename descclassname"><span class="pre">mango.models.pyomo.</span></span><span class="sig-name descname"><span class="pre">model_data_to_excel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'model_data.xlsx'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#mango.models.pyomo.model_data_to_excel" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a pyomo instance to excel.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model_data</strong> – dict with model data for pyomo model.</p></li>
<li><p><strong>path</strong> – path of the directory where to save the file</p></li>
<li><p><strong>clean</strong> – discard variables with values of 0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>nothing</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>

        </article>
        
        <div class="navigation flex print:hidden">
  <div class="navigation-prev">
    <a href="clients.html">
      <i class="i-icon chevron"></i>
      <div class="page-info">
        <span>Previous</span>
        
        <div class="title">Clients</div>
        
      </div>
    </a>
  </div>
  <div class="navigation-next">
    <a href="../bib.html">
      <div class="page-info">
        <span>Next</span>
        <div class="title">Bibliography</div>
      </div>
      <i class="i-icon chevron"></i>
    </a>
  </div>
</div>
        
      </div>
    </div>
  </main>
</div>
<footer class="sy-foot">
  <div class="sy-foot-inner sy-container mx-auto">
    
    <div class="sy-foot-reserved md:flex justify-between items-center">
      <div class="sy-foot-copyright"><p>2023, baobab soluciones</p>
        
        <p>
          Made with
          
          <a href="https://www.sphinx-doc.org/">Sphinx</a> and
          
          <a href="https://shibuya.lepture.com">Shibuya theme</a>.
        </p>
      </div>
      <div class="sy-foot-socials">
      </div>
    </div>
  </div>
</footer></div>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=158ece9d"></script>
      <script src="../_static/doctools.js?v=888ff710"></script>
      <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
      <script src="../_static/shibuya.js?v=ccb09f80"></script>
    
</body>
</html>